source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2075 @@
+    .socket_close_fn = s_socket_close_fn,
+    .socket_shutdown_dir_fn = s_socket_shutdown_dir_fn,
+    .socket_set_options_fn = s_socket_set_options_fn,
+    .socket_assign_to_event_loop_fn = s_socket_assign_to_event_loop_fn,
+    .socket_subscribe_to_readable_events_fn = s_socket_subscribe_to_readable_events_fn,
+    .socket_read_fn = s_socket_read_fn,
+    .socket_write_fn = s_socket_write_fn,
+    .socket_get_error_fn = s_socket_get_error_fn,
+    .socket_is_open_fn = s_socket_is_open_fn,
+    .socket_set_close_callback = s_set_close_callback,
+    .socket_set_cleanup_callback = s_set_cleanup_callback,
+};
+
+static void s_schedule_next_read(struct nw_socket *socket);
+
+static void s_socket_cleanup_fn(struct aws_socket *socket) {
+    if (!socket->impl) {
+        /* protect from double clean */
+        return;
+    }
+
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p nw_socket=%p: is cleanup...", (void *)socket, (void *)socket->impl);
+    if (aws_socket_is_open(socket)) {
+        AWS_LOGF_DEBUG(
+            AWS_LS_IO_SOCKET, "id=%p nw_socket=%p: is still open, closing...", (void *)socket, (void *)socket->impl);
+        aws_socket_close(socket);
+    }
+
+    struct nw_socket *nw_socket = socket->impl;
+
+    if (s_validate_event_loop(socket->event_loop) && !aws_event_loop_thread_is_callers_thread(socket->event_loop)) {
+        aws_mutex_lock(&nw_socket->synced_data.lock);
+        nw_socket->synced_data.base_socket = NULL;
+        aws_mutex_unlock(&nw_socket->synced_data.lock);
+    } else {
+        // If we are already on event loop or event loop is unavailable, we should already acquire the lock for base
+        // socket access
+        nw_socket->synced_data.base_socket = NULL;
+    }
+
+    aws_ref_count_release(&nw_socket->ref_count);
+    socket->impl = NULL;
+    AWS_ZERO_STRUCT(*socket);
+}
+
+struct read_queue_node {
+    struct aws_allocator *allocator;
+    dispatch_data_t received_data;
+    struct aws_linked_list_node node;
+    size_t current_offset;
+};
+
+static void s_clean_up_read_queue_node(struct read_queue_node *node) {
+    /* releases reference count on dispatch_data_t that was increased during creation of read_queue_node */
+    dispatch_release(node->received_data);
+    aws_mem_release(node->allocator, node);
+}
+
+struct socket_close_complete_args {
+    struct aws_allocator *allocator;
+    aws_socket_on_shutdown_complete_fn *shutdown_complete_fn;
+    void *user_data;
+    struct nw_socket *nw_socket;
+};
+
+static void s_close_complete_callback(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)status;
+    struct socket_close_complete_args *task_arg = arg;
+    struct aws_allocator *allocator = task_arg->allocator;
+    if (task_arg->shutdown_complete_fn) {
+        task_arg->shutdown_complete_fn(task_arg->user_data);
+    }
+    aws_ref_count_release(&task_arg->nw_socket->ref_count);
+    aws_mem_release(allocator, task_arg);
```

Comment:
Given that the task and the args are tightly coupled, why not just have the task be a member of socket_close_complete_args?

Reply:
I didnt see why that's necessary. The task is accessible from the task function call regardless.

Reply:
It's strictly simpler to allocate and free one thing than two.  As examples: 

* https://github.com/awslabs/aws-c-mqtt/blob/main/source/v5/mqtt5_client.c#L852-L858
* https://github.com/awslabs/aws-c-mqtt/blob/main/source/v5/mqtt5_client.c#L2101-L2107
* https://github.com/awslabs/aws-c-mqtt/blob/main/source/v5/mqtt5_to_mqtt3_adapter.c#L57-L61